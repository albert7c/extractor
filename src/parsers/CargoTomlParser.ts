import {Dependency, DependencyManagementFile} from "../../api/deps";
import {TomlParser} from "./Parser";

const organization = "__global__";
const scopes = [ "direct" ];

export default class CargoTomlParser extends TomlParser {
    public pathMatch(path: string): boolean {
        return super.pathMatch(path);
    }

    public parseToml(toml: any): DependencyManagementFile {

        const dependencies: Dependency[] = Object.keys(toml.dependencies)
            .map((name) => {
                const val = toml.dependencies[name];

                if (val instanceof String) {
                    return {
                        organization,
                        module: name,
                        versionConstraint: val,
                        scopes,
                    };
                } else {
                    return {
                        organization,
                        module: name,
                        versionConstraint: val.branch,
                        scopes,
                    };
                }
            });

        return {
            language: "rust",
            system: "cargo",
            organization,
            module: toml.package.name,
            version: toml.package.version,
            dependencies,
        };
    }
}
