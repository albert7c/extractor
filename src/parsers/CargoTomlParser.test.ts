import {readFile} from "fs";
import {promisify} from "util";
import CargoTomlParser from "./CargoTomlParser";

const readFileAsync = promisify(readFile);

describe("CargoTomlParser", () => {
    test("fullParse", async () => {
        const gopkgToml = require.resolve("./testdata/Cargo.toml");
        const buffer = await readFileAsync(gopkgToml);
        const content = buffer.toString();

        const parser = new CargoTomlParser();

        const actual = parser.parse(gopkgToml, content);

        expect(actual).toMatchSnapshot();
        expect(JSON.stringify(actual, null, 2)).toMatchSnapshot();
    });
});
