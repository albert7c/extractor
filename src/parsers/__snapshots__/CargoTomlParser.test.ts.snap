// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CargoTomlParser fullParse 1`] = `
Object {
  "dependencies": Array [
    Object {
      "module": "bytes",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "clap",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "futures",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "hyper",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "log",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "ordermap",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "pretty_env_logger",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "rand",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "rustls",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": "make_server_sni_public",
    },
    Object {
      "module": "serde",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "serde_derive",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "serde_json",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "serde_yaml",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "tacho",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "tokio-core",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "tokio-io",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "tokio-service",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "tokio-timer",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
    Object {
      "module": "url",
      "organization": "__global__",
      "scopes": Array [
        "direct",
      ],
      "versionConstraint": undefined,
    },
  ],
  "language": "rust",
  "module": "linkerd-tcp",
  "organization": "__global__",
  "system": "cargo",
  "version": "0.1.1",
}
`;

exports[`CargoTomlParser fullParse 2`] = `
"{
  \\"language\\": \\"rust\\",
  \\"system\\": \\"cargo\\",
  \\"organization\\": \\"__global__\\",
  \\"module\\": \\"linkerd-tcp\\",
  \\"version\\": \\"0.1.1\\",
  \\"dependencies\\": [
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"bytes\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"clap\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"futures\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"hyper\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"log\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"ordermap\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"pretty_env_logger\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"rand\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"rustls\\",
      \\"versionConstraint\\": \\"make_server_sni_public\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"serde\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"serde_derive\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"serde_json\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"serde_yaml\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"tacho\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"tokio-core\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"tokio-io\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"tokio-service\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"tokio-timer\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    },
    {
      \\"organization\\": \\"__global__\\",
      \\"module\\": \\"url\\",
      \\"scopes\\": [
        \\"direct\\"
      ]
    }
  ]
}"
`;
